Wrong line numbers reported under Python 3.8 with v0.2.x

Hello

Given the following code snippet
```python
import json


def another(x):
    """Another test function.

    This function is here to add extra lines to this snippet

    Parameters
    ----------
    x : int
        An arbitrary variable
    """
    json.loads(str(x))


# This is line 17
def f(id_):
    """Test function.

    Parameters
    ----------
    id_ : str. # This is line 23
        The identifier of the object
    """
    print("Hello world!")

```

I have the following flake8 configuration (nothing specific for rst-docstrings, but including for completeness' sake):
```
[flake8]
max_line_length = 120
max_complexity = 10
doctests = True
ignore = S101,E501,E203,W503,FS003
```

Python version:
`Python 3.8.8`

Then when I run the flake8 (with  `flake8-rst-docstrings==0.2.1`) on the following snippet I receive:
`/SCRATCHES/poc_flake8_docstrings.py:17:1: RST306 Unknown target name: "id".`

when I run the flake8 (with  `flake8-rst-docstrings==0.1.2`) on the following snippet I receive:
`/SCRATCHES/poc_flake8_docstrings.py:23:1: RST306 Unknown target name: "id".`

As you can observe, when using the latest version (happens for 0.2.0 as well), flake8 shows us the wrong line number.
Since I don't wan't to fix the doc, but ignore it. I now have to place a `# noqa: RST306` on line 17. But in my real code, I have my function decorated, so I have to put the noqa behind an argument in my decorator and that makes no sense at all.

The behaviour is actually wrong as well, but this is mentioned in https://github.com/peterjc/flake8-rst-docstrings/issues/26

If you need any more info, or if I can be of any more assistance, don't hesitate :)

Hmm. It is working for me:

```
$ flake8 bug.py
bug.py:1:1: D100 Missing docstring in public module
bug.py:23:1: RST306 Unknown target name: "id".
```

So, are we using the same things?

```
$ flake8 --version
3.8.1 (assertive: 1.2.1, flake8-blind-except: 0.1.1, flake8-bugbear: 20.11.1,
flake8-comprehensions: 3.2.2, flake8-docstrings: 1.5.0, pydocstyle: 5.0.2,
flake8-sfs: 0.0.3, flake8_implicit_str_concat: 0.1.0, mccabe: 0.6.1,
pycodestyle: 2.6.0, pyflakes: 2.2.0, rst-docstrings: 0.2.1) CPython 3.7.8 on
Darwin
```

This might be something in Python and/or flake8 from the AST parsing side...

OK, so this is not due to flake8 version 3.9.1 on Mac,

```
$ flake8 --version
3.9.1 (assertive: 1.2.1, flake8-blind-except: 0.1.1, flake8-bugbear: 20.11.1,
flake8-comprehensions: 3.2.2, flake8-docstrings: 1.5.0, pydocstyle: 5.0.2,
flake8-sfs: 0.0.3, flake8_implicit_str_concat: 0.1.0, mccabe: 0.6.1,
pycodestyle: 2.7.0, pyflakes: 2.3.1, rst-docstrings: 0.2.1) CPython 3.7.8 on
Darwin
$ flake8 bug.py 
bug.py:1:1: D100 Missing docstring in public module
bug.py:23:1: RST306 Unknown target name: "id".
```

It might be due to the Python version, given AST parsing is involved.

--- a/README.rst
+++ b/README.rst
@@ -236,6 +236,8 @@ Version History
 ======= ========== ===========================================================
 Version Released   Changes
 ------- ---------- -----------------------------------------------------------
+v0.2.2  2021-04-30 - Fixed line number problem under Python 3.8 or later.
+                   - Corrected off-by-one line number in module docstrings.
 v0.2.1  2021-04-23 - Minor internal style change.
 v0.2.0  2021-04-23 - Use AST from flake8, not re-parsing with pydocstyle.
                    - Drops ``RST901`` (internal problem with parser).

--- a/flake8_rst_docstrings.py
+++ b/flake8_rst_docstrings.py
@@ -175,10 +175,23 @@ class reStructuredTextChecker(object):
                     continue
 
                 if rst_errors:
-                    start = node.body[0].lineno - len(
-                        ast.get_docstring(node, clean=False).splitlines()
+                    try:
+                        node.body[0].end_lineno
+                        # Worked, on Python 3.8+ and can trust the start
+                        start = node.body[0].lineno - 1  # AST value 1 based
+                    except AttributeError:
+                        # On Python 3.7 or older, and must compute start line
+                        start = node.body[0].lineno - len(
+                            ast.get_docstring(node, clean=False).splitlines()
+                        )
+                        if isinstance(node, ast.Module):
+                            start -= 1  # Why?
+                    assert (
+                        node.body[0].lineno >= 1 and start >= 0
+                    ), "Bad start line, node line number %i for: %s\n" % (
+                        node.body[0].lineno,
+                        docstring,
                     )
-
                 for rst_error in rst_errors:
                     # TODO - make this a configuration option?
                     if rst_error.level <= 1:

--- a/flake8_rst_docstrings.py
+++ b/flake8_rst_docstrings.py
@@ -9,7 +9,7 @@ import ast
 import restructuredtext_lint as rst_lint
 
 
-__version__ = "0.2.1"
+__version__ = "0.2.2"
 
 
 rst_prefix = "RST"

--- a/tests/RST202/bullets.py
+++ b/tests/RST202/bullets.py
@@ -9,7 +9,7 @@ Other fruit.
 It should therefore fail RST validation::
 
     $ flake8 --select RST RST202/bullets.py
-    RST202/bullets.py:8:1: RST202 Bullet list ends without a blank line; unexpected unindent.
+    RST202/bullets.py:7:1: RST202 Bullet list ends without a blank line; unexpected unindent.
     RST202/bullets.py:38:1: RST202 Bullet list ends without a blank line; unexpected unindent.
 
 The end.

--- a/tests/RST206/field_lists.py
+++ b/tests/RST206/field_lists.py
@@ -21,7 +21,7 @@ Because the line above is not indented or separated from
 the field list by a blank line, it is not valid RST:
 
     $ flake8 --select RST RST206/fields_list.py
-    RST206/field_lists.py:19:1: RST206 Field list ends without a blank line; unexpected unindent.
+    RST206/field_lists.py:18:1: RST206 Field list ends without a blank line; unexpected unindent.
 
 The end.
 """  # noqa: E501

--- a/tests/RST208/option_list.py
+++ b/tests/RST208/option_list.py
@@ -29,7 +29,7 @@ There ought to be a blank line before that final comment,
 there isn't so this fails validation:
 
     $ flake8 --select RST RST208/option_list.py
-    RST208/option_list.py:27:1: RST208 Option list ends without a blank line; unexpected unindent.
+    RST208/option_list.py:26:1: RST208 Option list ends without a blank line; unexpected unindent.
 
 The end.
 """  # noqa: E501

--- a/tests/RST210/strong.py
+++ b/tests/RST210/strong.py
@@ -8,7 +8,7 @@ Here **strong is missing a closing double asterisk.
 That is considered to be an error, and should fail::
 
     $ flake8 --select RST RST210/strong.py
-    RST210/strong.py:7:1: RST210 Inline strong start-string without end-string.
+    RST210/strong.py:6:1: RST210 Inline strong start-string without end-string.
 
 """
 

--- a/tests/RST212/short_underline.py
+++ b/tests/RST212/short_underline.py
@@ -11,7 +11,7 @@ There is a missing equals sign on the above underline,
 and that is considered an error. This should fail::
 
     $ flake8 --select RST RST212/short_underline.py
-    RST212/short_underline.py:10:1: RST212 Title underline too short.
+    RST212/short_underline.py:9:1: RST212 Title underline too short.
 
 Nice
 ====

--- a/tests/RST213/emphasis.py
+++ b/tests/RST213/emphasis.py
@@ -8,7 +8,7 @@ Here *emphasis is missing a closing asterisk.
 That is considered to be an error, and should fail::
 
     $ flake8 --select RST RST213/emphasis.py
-    RST213/emphasis.py:7:1: RST213 Inline emphasis start-string without end-string.
+    RST213/emphasis.py:6:1: RST213 Inline emphasis start-string without end-string.
 
 """
 

--- a/tests/RST214/literal.py
+++ b/tests/RST214/literal.py
@@ -8,7 +8,7 @@ Here ``literal is missing the closing backticks.
 That is considered to be an error, and should fail::
 
     $ flake8 --select RST RST214/literal.py
-    RST214/literal.py:7:1: RST214 Inline emphasis start-string without end-string.
+    RST214/literal.py:6:1: RST214 Inline emphasis start-string without end-string.
 
 """
 

--- a/tests/RST215/backticks.py
+++ b/tests/RST215/backticks.py
@@ -8,7 +8,7 @@ Here `example is missing a closing backtick.
 That is considered to be an error, and should fail::
 
     $ flake8 --select RST RST215/backticks.py
-    RST215/backticks.py:7:1: RST215 Inline interpreted text or phrase reference start-string without end-string.
+    RST215/backticks.py:6:1: RST215 Inline interpreted text or phrase reference start-string without end-string.
 
 """  # noqa: E501
 

--- a/tests/RST217/roles.py
+++ b/tests/RST217/roles.py
@@ -10,7 +10,7 @@ However, trailing underscores have special meaning for referencing, thus
 `code`:example:_ is considered to be an error:
 
     $ flake8 --select RST RST217/roles.py
-    RST217/roles.py:10:1: RST17 Mismatch: both interpreted text role suffix and reference suffix.
+    RST217/roles.py:9:1: RST17 Mismatch: both interpreted text role suffix and reference suffix.
 
 """  # noqa: E501
 

--- a/tests/RST218/no_literal_block.py
+++ b/tests/RST218/no_literal_block.py
@@ -5,7 +5,7 @@ indented text is a literal block. The following
 code snippet is a typical usage::
 
     $ flake8 --select RST RST218/no_literal_block.py
-    RST218/no_literal_block.py:14:1: RST218 Literal block expected; none found.
+    RST218/no_literal_block.py:13:1: RST218 Literal block expected; none found.
 
 This file triggers an error because this paragraph
 says there should be a following literal block, but

--- a/tests/RST305/subsitution.py
+++ b/tests/RST305/subsitution.py
@@ -12,7 +12,7 @@ Here the |bar| substituion definition is missing, so this
 docstring in isolation should fail validation::
 
     $ flake8 --select RST RST305/substitution.py
-    RST305/subsitution.py:12:1: RST305 Undefined substitution referenced: "bar"
+    RST305/subsitution.py:11:1: RST305 Undefined substitution referenced: "bar"
 
 """
 

--- a/tests/RST306/unknown_target.py
+++ b/tests/RST306/unknown_target.py
@@ -12,7 +12,7 @@ Here a missing-link_ hyperlink is used, so this docstring in
 isolation should fail validation::
 
     $ flake8 --select RST  RST306/unknown_target.py
-    RST306/unknown_target.py:12:1: RST306 Unknown target name: "missing-link".
+    RST306/unknown_target.py:11:1: RST306 Unknown target name: "missing-link".
 
 """
 

--- a/tests/RST307/code_invalid_arg.py
+++ b/tests/RST307/code_invalid_arg.py
@@ -35,9 +35,9 @@ And:
 Sadly docutils considers all three examples to be invalid:
 
     $ flake8 --select RST  RST307/code_invalid_arg.py
-    RST307/code_invalid_arg.py:14:1: RST307 Error in "code" directive
-    RST307/code_invalid_arg.py:23:1: RST307 Error in "code-block" directive:
-    RST307/code_invalid_arg.py:31:1: RST307 Error in "code-block" directive:
+    RST307/code_invalid_arg.py:13:1: RST307 Error in "code" directive
+    RST307/code_invalid_arg.py:22:1: RST307 Error in "code-block" directive:
+    RST307/code_invalid_arg.py:30:1: RST307 Error in "code-block" directive:
 
 """
 

